### PR TITLE
137 constant change

### PR DESCRIPTION
**Pull request by rsantellan**
## Description
Externalize the colors of the message to table. #137 

## Checklist

- [X] Commit messages are descriptive and explain the rationale for changes
- [X] I used the same coding style as the surrounding code 
- [X] I have tested the proposed changes
- [X] I extended the documentation, if necessary
- [ ] The proposed change builds also on our build bots (check after submission)

### CODE REVIEW

**Review by bradallred — CHANGES_REQUESTED** [review]

File: gemrb/core/DisplayMessage.h [hunk 1]
```diff
@@ -36,13 +36,7 @@
...
 namespace GemRB {
 
-static constexpr Color DMC_WHITE {0xf0, 0xf0, 0xf0, 0xff};
-static constexpr Color DMC_RED = ColorRed;
-static constexpr Color DMC_LIGHTGREY {0xd7, 0xd7, 0xbe, 0xff};
-static constexpr Color DMC_BG2XPGREEN {0xbc, 0xef, 0xbc, 0xff};
-static constexpr Color DMC_GOLD {0xc0, 0xc0, 0x00, 0xff};
-static constexpr Color DMC_DIALOG = DMC_WHITE;
-static constexpr Color DMC_DIALOGPARTY = {0x80, 0x80, 0xff, 0xff};
+enum { DMC_DIALOG, DMC_DIALOGPARTY, DMC_GOLD, DMC_RED, DMC_BG2XPGREEN, DMC_LIGHTGREY, DMC_WHITE };
```
[the remark this answers is not in the document]
Comment (lynxlynxlynx) [reply gemrb/core/DisplayMessage.h:39]: That's the plan, but I made him go slower to be clearer and to avoid conflicts with the utf-16 branch.

**Comment by lynxlynxlynx** [comment]
Some of the DisplayMessage functions are now redundant. For example there's no need for all that indirection with DisplayRollStringName, since it never uses dynamic colors. So just change the original function, there's no need for a wrapper. Then do the same with any other such function.

File: gemrb/unhardcoded/bg1/colors.2da [hunk 2]
```diff
@@ -7,3 +7,10 @@ MAPICNBG		0xf0b080ff
 MAPTXTBG		0xf0b080ff
 ALTDOOR			0xff00ffff
 HOVERDOOR		0xff00ffff
+WHITE			0xf0f0f0ff
+RED				0xff0000ff
+LIGHT_GREY		0xd7d7beff
+GOT_XP			0xbcefbcff
+GOT_GOLD		0xc0c000ff
+DIALOG			0xf0f0f0ff
+DIALOG_PARTY	0x8080ffff
```
[the remark this answers is not in the document]
Comment (rsantellan) [reply gemrb/unhardcoded/bg1/colors.2da:16]: I have change this to add the new line and better formatting.

File: gemrb/core/DisplayMessage.h [hunk 3]
```diff
@@ -64,10 +58,22 @@ class GEM_EXPORT DisplayMessage
 	static String ResolveStringRef(ieStrRef);
 	
 	template<typename ...ARGS>
-	void DisplayRollStringName(const String& fmt, const Color &color, const Scriptable *speaker, ARGS&& ...args) const {
+	void DisplayRollStringName(const String& fmt, unsigned char color, const Scriptable *speaker, ARGS&& ...args) const {
 		String formatted = fmt::format(fmt, std::forward<ARGS>(args)...);
 		DisplayStringName(std::move(formatted), color, speaker);
 	}
+	/** displays a string constant in the textarea */
+	void DisplayConstantString(size_t stridx, const Color &color, Scriptable *speaker=NULL) const;
```
[the remark this answers is not in the document]
Comment (rsantellan) [reply gemrb/core/DisplayMessage.h:66]: Change to nullptr instead of NULL

**Comment by rsantellan** [comment]
> Some of the DisplayMessage functions are now redundant. For example there's no need for all that indirection with DisplayRollStringName, since it never uses dynamic colors. So just change the original function, there's no need for a wrapper. Then do the same with any other such function.

The DisplayRollStringName was like that, I don't really understand it fully so I didn't touch it. Will check which of the DisplayMessage are redundant.

**Comment by rsantellan** [comment]
> > Some of the DisplayMessage functions are now redundant. For example there's no need for all that indirection with DisplayRollStringName, since it never uses dynamic colors. So just change the original function, there's no need for a wrapper. Then do the same with any other such function.
> 
> The DisplayRollStringName was like that, I don't really understand it fully so I didn't touch it. Will check which of the DisplayMessage are redundant.

I'm a little stuck with this part. Don't know how to improve it.

**Comment by lynxlynxlynx** [comment]
Is anyone still calling
`void DisplayRollStringName(const String& fmt, unsigned char color, const Scriptable *speaker, ARGS&& ...args) const {` ? You should be able to merge it with the other function.

**Comment by rsantellan** [comment]
> 
```
grep 'DisplayRollStringName' * -nr
gemrb/includes/ie_types.h:77:	// params to DisplayRollStringName
gemrb/core/GameScript/Actions.cpp:4562:		displaymsg->DisplayRollStringName(ieStrRef::ROLL12, DMC_LIGHTGREY, snd, 10+skill, roll+level, wismod);
gemrb/core/Scriptable/Container.cpp:193:			displaymsg->DisplayRollStringName(ieStrRef::ROLL11, DMC_LIGHTGREY, actor, stat-dexmod, LockDifficulty, dexmod);
gemrb/core/Scriptable/Container.cpp:232:		displaymsg->DisplayRollStringName(ieStrRef::ROLL1, DMC_LIGHTGREY, actor, roll, bonus, LockDifficulty);
gemrb/core/Scriptable/Scriptable.cpp:1147:		displaymsg->DisplayRollStringName(ieStrRef::ROLL21, DMC_LIGHTGREY, actor, roll, chance);
gemrb/core/Scriptable/Scriptable.cpp:1194:			displaymsg->DisplayRollStringName(ieStrRef::ROLL15, DMC_LIGHTGREY, detective, Spellcraft+IntMod, AdjustedSpellLevel, IntMod);
gemrb/core/Scriptable/Scriptable.cpp:1862:			displaymsg->DisplayRollStringName(ieStrRef::ROLL13, DMC_LIGHTGREY, detective, skill-bonus, TrapDetectionDiff, bonus);
gemrb/core/Scriptable/Actor.cpp:3626:		displaymsg->DisplayRollStringName(ieStrRef::ROLL22, DMC_LIGHTGREY, this, roll, save, modifier, spellLevel, saveBonus);
gemrb/core/Scriptable/Actor.cpp:3630:		displaymsg->DisplayRollStringName(ieStrRef::ROLL23, DMC_LIGHTGREY, this, roll, save, modifier, spellLevel, saveBonus);
gemrb/core/Scriptable/Actor.cpp:4363:		displaymsg->DisplayRollStringName(ieStrRef::ROLL19, DMC_LIGHTGREY, this, roll, concentration, bonus, damage, spellLevel);
gemrb/core/Scriptable/Actor.cpp:4415:			displaymsg->DisplayRollStringName(ieStrRef::ROLL20, DMC_LIGHTGREY, this, ToHit.GetTotal());
gemrb/core/Scriptable/Actor.cpp:9235:	displaymsg->DisplayRollStringName(ieStrRef::ROLL14, DMC_LIGHTGREY, this, skill + roll, level + 20, level);
gemrb/core/Scriptable/Actor.cpp:9399:			displaymsg->DisplayRollStringName(ieStrRef::ROLL18, DMC_LIGHTGREY, this, sneakAttackDamage);
gemrb/core/Scriptable/Actor.cpp:9404:			displaymsg->DisplayRollStringName(ieStrRef::ROLL17, DMC_LIGHTGREY, this, sneakAttackDamage);
gemrb/core/Scriptable/Actor.cpp:9418:		//displaymsg->DisplayRollStringName(25053, DMC_LIGHTGREY, this, extraDamage);
gemrb/core/Scriptable/Actor.cpp:10663:			displaymsg->DisplayRollStringName(ieStrRef::ROLL10, DMC_LIGHTGREY, actor, skill-bonus, roll, bonus);
gemrb/core/Scriptable/Actor.cpp:10667:			displaymsg->DisplayRollStringName(ieStrRef::ROLL8, DMC_LIGHTGREY, actor, skill, targetDC, roll);
gemrb/core/Scriptable/Actor.cpp:10671:			displaymsg->DisplayRollStringName(ieStrRef::ROLL7, DMC_LIGHTGREY, actor, skill, targetDC, roll);
gemrb/core/Scriptable/Actor.cpp:10772:	displaymsg->DisplayRollStringName(ieStrRef::ROLL9, DMC_LIGHTGREY, this, skill/7, roll, GetAbilityBonus(IE_DEX));
gemrb/core/Scriptable/Actor.cpp:10810:			displaymsg->DisplayRollStringName(ieStrRef::ROLL2, DMC_LIGHTGREY, this, skill, targetDC, roll);
gemrb/core/Scriptable/Actor.cpp:10837:			displaymsg->DisplayRollStringName(ieStrRef::ROLL0, DMC_LIGHTGREY, this, skill, targetDC, roll);
gemrb/core/Scriptable/Actor.cpp:11201:			displaymsg->DisplayRollStringName(ieStrRef::ROLL4, DMC_LIGHTGREY, this, roll + concentration, 15 + spellLevel, bonus);
gemrb/core/Scriptable/Actor.cpp:11203:			displaymsg->DisplayRollStringName(ieStrRef::ROLL5, DMC_LIGHTGREY, this);
gemrb/core/Scriptable/Actor.cpp:11209:			displaymsg->DisplayRollStringName(ieStrRef::ROLL3, DMC_LIGHTGREY, this, roll + concentration, 15 + spellLevel, bonus);
gemrb/core/Scriptable/Door.cpp:351:			displaymsg->DisplayRollStringName(ieStrRef::ROLL6, DMC_LIGHTGREY, actor, roll, skill-bonus, bonus, trapDC);
gemrb/core/Scriptable/Door.cpp:363:			displaymsg->DisplayRollStringName(ieStrRef::ROLL6, DMC_LIGHTGREY, actor, roll, skill-bonus, bonus, trapDC);
gemrb/core/Scriptable/Door.cpp:390:			displaymsg->DisplayRollStringName(ieStrRef::ROLL11, DMC_LIGHTGREY, actor, stat-dexmod, LockDifficulty, dexmod);
gemrb/core/Scriptable/Door.cpp:429:		displaymsg->DisplayRollStringName(ieStrRef::ROLL1, DMC_LIGHTGREY, actor, roll, bonus, LockDifficulty);
gemrb/core/DisplayMessage.h:61:	void DisplayRollStringName(const String& fmt, unsigned char color, const Scriptable *speaker, ARGS&& ...args) const {
gemrb/core/DisplayMessage.h:112:	void DisplayRollStringName(ieStrRef stridx, unsigned char color, const Scriptable *speaker, ARGS&& ...args) const {
gemrb/core/DisplayMessage.h:115:			DisplayRollStringName(fmt, color, speaker, std::forward<ARGS>(args)...);
gemrb/core/EffectQueue.cpp:1020:		displaymsg->DisplayRollStringName(ieStrRef::ROLL16, DMC_LIGHTGREY, actor, val, roll, fx->CasterLevel, penetration);
```
I have does, the idea is change that for this: **DisplayStringName** ? What I do with the extra arguments ?

**Comment by lynxlynxlynx** [comment]
I don't see any user there, besides the one it should be merged with — all take a ieStrRef.

btw, `git grep` is very handy for grepping in git trees.

**Comment by rsantellan** [comment]
> I don't see any user there, besides the one it should be merged with — all take a ieStrRef.
> 
> btw, `git grep` is very handy for grepping in git trees.

Remove the private one and use only the public one, the thing that really complicates me is the **String formatted = fmt::format(fmt, std::forward<ARGS>(args)...);** don't know how to change it to call to other functions.

**Comment by lynxlynxlynx** [comment]
The last commit looks good. 👌 There's no need for that line to call any other function, it's basically a printf call.

**Comment by lynxlynxlynx** [comment]
Now you can check if any other function can be removed, since you added many new ones earlier.

**Comment by lynxlynxlynx** [comment]
If I had to guess, I'd say that's a consequence of displaymsg's external linkage with no visibility limits.

**Comment by rsantellan** [comment]
> If I had to guess, I'd say that's a consequence of displaymsg's external linkage with no visibility limits.

I double check all the functions and was able to remove a few of them, but a lot of them didn't want to change them because was used in other parts of the code.

File: gemrb/core/DisplayMessage.h [hunk 4]
```diff
@@ -63,46 +57,49 @@ class GEM_EXPORT DisplayMessage
...
+	void DisplayStringName(String text, unsigned char color, const Scriptable *speaker) const;
+	void DisplayStringName(String text, const Color &color, const Scriptable *speaker) const;
+	void DisplayStringName(ieStrRef stridx, const Color &color, const Scriptable *speaker, STRING_FLAGS flags) const;
+	/** displays a string constant in the textarea, starting with speaker's name, also replaces one numeric value (it is a format string) */
+	void DisplayConstantStringNameValue(size_t stridx, unsigned char color, const Scriptable *speaker, int value) const;
+	/** displays actor name - action : parameter */
+	void DisplayConstantStringNameString(size_t stridx, unsigned char color, size_t stridx2, const Scriptable *actor) const;
 	/** displays a string in the textarea */
 	void DisplayString(const String& text) const;
+	void DisplayString(ieStrRef stridx, unsigned char color, STRING_FLAGS flags) const;
```
Comment (lynxlynxlynx) [inline gemrb/core/DisplayMessage.h:91]: If you resolve the strref in GemRB_DisplayString, this one can go. Same for the variant of DisplayStringName that it is using.

File: gemrb/core/DisplayMessage.h [hunk 5]
```diff
@@ -63,46 +57,49 @@ class GEM_EXPORT DisplayMessage
...
+	void DisplayString(const String& text, unsigned char color, Scriptable *target) const;
 	void DisplayString(ieStrRef stridx, const Color &color, STRING_FLAGS flags) const;
 	void DisplayString(String text, const Color &color, Scriptable *target) const;
```
Comment (lynxlynxlynx) [inline gemrb/core/DisplayMessage.h:94]: This one can be private now.

**Comment by rsantellan** [comment]
@lynxlynxlynx I did a merge of the master to fix the problems but I believe now it was a mistake... Let me know how to improve it.

**Comment by lynxlynxlynx** [comment]
Oh no, it's indeed a mess. Next time use a rebase (eg. git pull --rebase) to avoid this, I'm not sure if it's trivially solvable. You could try rebasing now, but I don't know if it would work.

What should either way is creating a new branch and cherry picking you work on top, commit by commit. Something like:
```bash
# - assuming your local "master" branch is up to date with our upstream one
# 1. filter out your own commits; this should print the commit SHAs that are not yet on master
git cherry master

# 2. create a fresh branch from master
git checkout -b 137-colors master

# 3. start copying the work over
# run this for each commit found in 1.
git cherry-pick the-commit-sha

# 4. testing

# 5. open a fresh PR with the new branch

# 6. delete the old branch
# git branch -D 137-constant-change
```

**Comment by rsantellan** [comment]
I have created a new PR https://github.com/gemrb/gemrb/pull/1575

I did it by hand because the rebase was a total mess.

**Pull request closed without merging**